runtime/caches/fileSystem.ts

```diff
@@ -0,0 +1,216 @@
+              cacheKey,
+            },
+          });
+          try {
+            const startTime = performance.now();
+            const data = await getFile(cacheKey);
+            const downloadDurationTime = performance.now() - startTime;
+
+            span.addEvent("file-system-get-data");
+
+            if (data === null) {
+              span.addEvent("cache-miss");
+              return undefined;
+            }
+            span.addEvent("cache-hit");
+
+            downloadDuration.record(downloadDurationTime, {
+              bufferSize: data.length,
+            });
+
+            return new Response(
+              data,
+            );
+          } catch (err) {
+            span.recordException(err);
+            throw err;
+          } finally {
+            span.end();
+          }
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/matchAll) */
+        matchAll: (
+          _request?: RequestInfo | URL,
+          _options?: CacheQueryOptions,
+        ): Promise<ReadonlyArray<Response>> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/put) */
+        put: async (
+          request: RequestInfo | URL,
+          response: Response,
+        ): Promise<void> => {
+          const req = new Request(request);
+          assertCanBeCached(req, response);
+
+          if (!response.body) {
+            return;
+          }
+
+          const cacheKey = await requestURLSHA1(request);
+          const buffer = await response.arrayBuffer()
+            .then((buffer) => new Uint8Array(buffer))
+            .then((buffer) => {
+              bufferSizeSumObserver.add(buffer.length);
+              return buffer;
+            });
+
+          const span = tracer.startSpan("file-system-put", {
+            attributes: {
+              cacheKey,
+            },
+          });
+
+          try {
+            try {
+              const setSpan = tracer.startSpan("file-system-set", {
+                attributes: { cacheKey },
+              });
+              putFile(cacheKey, buffer).catch(
+                (err) => {
+                  console.error("file system error", err);
+                  setSpan.recordException(err);
+                },
+              ).finally(() => {
+                setSpan.end();
+              }); // do not await for setting cache
+            } catch (error) {
+              logger.error(`error saving to file system ${error?.message}`);
+            }
+          } catch (err) {
+            span.recordException(err);
+            throw err;
+          } finally {
+            span.end();
+          }
+        },
+      });
+    },
+  };
+
+  // Check if the cache directory exists, if not, create it
+  if (FILE_SYSTEM_CACHE_DIRECTORY && !existsSync(FILE_SYSTEM_CACHE_DIRECTORY)) {
+    Deno.mkdirSync(FILE_SYSTEM_CACHE_DIRECTORY, { recursive: true });
+  }
+  return caches;
```

Comment:
we can't do that at module-level. You need to do it lazy, also, we should have a way to check if we have write permissions and then fallback to undefined

Reply:
Done